fat: Parse BIOS Parameters

kallisti5 opened the pull request:
* Eventually we can read these from the image
  instead of guessing via the const's at the
  top of image.rs

kallisti5 committed:
fat: Parse BIOS Parameters

* Eventually we can read these from the image
  instead of guessing via the const's at the
  top of image.rs

kallisti5 commented:
kallisti5@nyx fatr :) $ ./target/debug/fatr ls ~/fat12.img 
BIOSParam { bytes_per_sector: 512, sectors_per_cluster: 4, reserved_sectors: 512, fat_count: 0, max_roots: 8192, sectors: 1784, media_id: 0, sectors_per_fat: 0 }

kallisti5@nyx fatr :) $ ./target/debug/fatr ls ~/fat16.img 
BIOSParam { bytes_per_sector: 512, sectors_per_cluster: 4, reserved_sectors: 512, fat_count: 0, max_roots: 0, sectors: 16632, media_id: 0, sectors_per_fat: 0 }

Example output. Now that we know the bytes_per_sector, fatr could start calculating those sector sizes more dynamically.

I'm not sure how far you want this project to go, but this is the only native Rust FAT create out there :-)

zanesterling commented on src/commands/list.rs:
```diff
@@ -10,6 +10,10 @@ pub fn list_files(args: &[String])
 
+    let params = image.bios_parameter();
+    // Just dump for testing. Remove when tried of me :-)
+    println!("{:?}", params);
```
Maybe hide behind a --verbose flag? [Clap](https://github.com/kbknapp/clap-rs) is a pretty excellent command-line parser. Or only compile in in debug builds.

kallisti5 replied:
Works for me. I was mostly dumping to show functionality since it isn't used for anything yet :-)  Will fix.

zanesterling commented on src/fat/bios_param.rs:
```diff
@@ -0,0 +1,30 @@
+	pub sectors_per_fat: u16,
+}
+
+/*
```
Avoid leaving commented code chunks lying around.

kallisti5 replied:
Sure. Will complete.  I planned on adding some convenience calculation functions to the bios_param code but didn't get around to it before submitting the PR.  I'll add it in or remove commented code.

zanesterling commented on src/fat/image.rs:
```diff
@@ -73,6 +74,19 @@ impl Image {
+            let p = &mut param as *mut BIOSParam as *mut [u8; 13];
+            ::std::ptr::write(p, bios_bytes);
+            param
+        };
```
Is this defined behavior? The compiler might reorder the struct's fields. As annoying as it is, prob should be done by hand.

kallisti5 replied:
The #[repr(C)] 's around the structs guarantee the struct's wont be reordered.

We aren't interested in the first few bytes (boot code), so this let us skip to the "good stuff" in boot_sector without needing to hack it up.

Definitely open to better solutions, after an hour or so talking to #rust in irc, this was their best solution :-)  I noticed the field seemed to be misaligned... but that behavior only occurred in one image, so I was hoping something in the image was wrong.

zanesterling reviewed (changes requested):
Good stuff. The unsafe piece seems sketch, though. See comments.

zanesterling commented:
I probably won't actively work on this crate. I built it for a one-off operating systems project I was working on a while back. I'd be happy to add you as a collab on the repo and an owner on the crate, though, if you want to tinker with it 😄

kallisti5 commented:
Heh, i'd be happy with that.  I'd still like some code review for as long as you have the free time since this is your project.   Given how cleanly this was written, i'd bet (basic) FAT16 support as well as FAT12 wouldn't be impossible.

kallisti5 committed:
bios_param: Further refine

* Add convience functions
* Add tests for bios_param
* Add some light rust documentation

kallisti5 committed:
bios_param: Remove unsafe code, leverage ByteOrder

* Handle 0-sector extensions in FAT16 and FAT32
* BIOSParam is no longer aligned to the binary format
* If we ever want to write to BIOSParam, we are going to need
  a correct binary representation of the data beyond BIOSParam.

kallisti5 commented:
Ok. Everything should be covered in this latest commit. I've removed the unsafe code as well and switched to the ByteOrder crate.

```
kallisti5@nyx fat :( $ ../../target/debug/fatr ls ~/fat12.img 
BIOSParam { bytes_per_sector: 512, sectors_per_cluster: 4, reserved_sectors: 1, fat_count: 2, max_roots: 512, sectors: 8192, media_id: 248, sectors_per_fat: 6 }
kallisti5@nyx fat :) $ ../../target/debug/fatr ls ~/fat16.img 
BIOSParam { bytes_per_sector: 512, sectors_per_cluster: 4, reserved_sectors: 4, fat_count: 2, max_roots: 512, sectors: 65536, media_id: 248, sectors_per_fat: 64 }
kallisti5@nyx fat :) $ ../../target/debug/fatr ls ~/fat32.img 
BIOSParam { bytes_per_sector: 512, sectors_per_cluster: 1, reserved_sectors: 32, fat_count: 2, max_roots: 0, sectors: 524288, media_id: 248, sectors_per_fat: 4033 }
```

zanesterling reviewed (approved):
LGTM
